File: JAVA/Unique Binary Search Trees II.java
public class Solution {
    public ArrayList<TreeNode> generateTrees(int n) {
        return generateTrees(1, n);
    }
     
    private ArrayList<TreeNode> generateTrees(int left, int right)
    {
        ArrayList<TreeNode> ans = new ArrayList<TreeNode>();
        if (left > right)
        {
            ans.add(null);
            return ans;
        }
        for (int i = left; i <= right; i++)
        {
            ArrayList<TreeNode> lefts = generateTrees(left, i-1);
            ArrayList<TreeNode> rights = generateTrees(i+1, right);
            for (int j = 0; j < lefts.size(); j++)
            {
                for (int k = 0; k < rights.size(); k++)
                {
                    TreeNode root = new TreeNode(i);
                    root.left = lefts.get(j);
                    root.right = rights.get(k);
                    ans.add(root);
                }
            }
        }
        return ans;
    }
}

File: Unique Binary Search Trees II.cpp
class Solution {
public:
    vector<TreeNode *> generateTrees(int n) {
        return generateTreesRe(1, n);
    }
    
    vector<TreeNode *> generateTreesRe(int x, int y) {
        vector<TreeNode *> result;
        if (x > y) {
            result.push_back(NULL);
        } else {
            for (int k = x; k <= y; k++) {
                vector<TreeNode*> left = generateTreesRe(x, k - 1);
                vector<TreeNode*> right = generateTreesRe(k + 1, y);
                for (int i = 0; i < left.size(); i++) {
                    for (int j = 0; j < right.size(); j++) {
                        TreeNode *root = new TreeNode(k);
                        root->left = left[i];
                        root->right = right[j];
                        result.push_back(root);
                    }
                }
            }
        }
        return result;
    }
};
